### ISSUE TITLE
JupyterLab

### ISSUE THREAD

**Issue by gabrielmagno** [catppuccin/catppuccin#2152]
### Link to repository

https://github.com/gabrielmagno/jupyterlab

### Screenshots

![latte](https://github.com/catppuccin/catppuccin/assets/244890/9c516c70-629f-4ebe-bfac-304ecbbc9ef4)
![frappe](https://github.com/catppuccin/catppuccin/assets/244890/8d8d38cb-27d8-44b6-be0b-affb3002aa59)
![macchiato](https://github.com/catppuccin/catppuccin/assets/244890/09520ece-63bc-4e04-a37d-b53b96854eb5)
![mocha](https://github.com/catppuccin/catppuccin/assets/244890/ff2f825b-57ee-4ab4-8d3c-199ef417c8a5)

### Any additional comments?

I have published a package called `catppuccin-jupyterlab` in PyPI:
https://pypi.org/project/catppuccin-jupyterlab/

So that users can install it with:
```
pip install catppuccin-jupyterlab
```

### Submission Guidelines

- [X] I have read and followed Catppuccin's submission guidelines.

**Comment by backwardspy** [comment]
hey, thanks for your contribution! i've tested it and noticed just a couple of issues.

<hr>

(click to expand each point)

<details>
<summary>when using a dark theme variant, dropdowns in the settings menu use a light background</summary>

![image](https://github.com/catppuccin/catppuccin/assets/289746/548b6b0e-ff37-4ddf-b45d-2351e91096b5)
however this appears to be an issue on the stock dark theme too, so it may not be fixable.

</details>

<details>
<summary>buttons look too light compared to other applications</summary>

![image](https://github.com/catppuccin/catppuccin/assets/289746/0df962e0-b57e-4316-a680-029e8876f270)

comparison with the jetbrains port for example:

![image](https://github.com/catppuccin/catppuccin/assets/289746/ba9990e3-66f0-4c91-b8bd-a941747f279c)

we generally recommend the `surface` colours for elements like buttons, with accent used as the primary option (where applicable.)
</details>

<hr>

the repo itself looks good, my only comment there is in the readme there's a typo in the note about accent colour: "rarelly" → "rarely". github also recently added the ability to use "alerts" to format things like notes and warnings:

```
> [!NOTE]
> Besides the "brand" color it is also possible to set the "accent" color (default is green), but it rarely appears in the JupyterLab interface.
```

becomes

> [!NOTE]
> Besides the "brand" color it is also possible to set the "accent" color (default is green), but it rarely appears in the JupyterLab interface.

<hr>

let me know if you have any questions or concerns about any of that. great work on the port, it genuinely looks very nice!

**Comment by gabrielmagno** [comment]
Thanks for the review and the insightful comments @backwardspy !

I will work on the modifications and notify here when I have a new version for review.

Regarding the term "brand color" I'm not sure if this is a broad term used externally by JupyterLab community (I actually don't remember seeing a theme that has the feature of personalization like this). What I can say is that "brand color" is at least an _internal_ term on the code. These are the internal CSS variables that we have to set:

```
--jp-brand-color0
--jp-brand-color1
--jp-brand-color2
--jp-brand-color3
--jp-accent-color0
--jp-accent-color1
--jp-accent-color2
--jp-accent-color3
```

As you can see, there is both "brand" color, and "accent" color. The "brand" color is the one that we see in most places in the interface. The "accent" color as a matter of fact I was not able to understand _where_ it is used in the interface. But since it has this name, I determined an opinionated default (green) and allowed the user to personalize it, just in case.

One possibility would be to expose only the "brand" color, but call it "accent" externally. Not sure if this could make confusion for someone more familiar with the JupyterLab colors.

That being said, I am totally open for suggestions, I have no strong attachment to my idea of exposing both colors, and not sure if that is the best idea.

What do you think?

**Comment by backwardspy** [comment]
that makes sense, thanks for the additional context!

i don't think it's a big deal. having them both be customisable is great and your readme explains the distinction well enough.

without knowing exactly how the accent-* variables are used it's hard to say if we should just make them both be the same colour, so i say keep it the way it is. the fact that they're named that way internally seems like a perfectly good reason to expose them like you have.

**Comment by gabrielmagno** [comment]
Hey @backwardspy .

Let me go trough each one of your comments and suggestions:

<details>
<summary>the repo itself looks good, my only comment there is in the readme there's a typo in the note about accent colour: "rarelly" → "rarely". github also recently added the ability to use "alerts" to format things like notes and warnings:</summary>
</br>
Done!
</details>

<details>
<summary>buttons look too light compared to other applications</summary>
Done!

I've made the "gray" buttons (internally called "reject button") darker, I'm using `surface0` now.

I also am using the brand color for "primary" buttons.

There are 3 button types in JupyterLab:
- Accept: traditionally blue. I was using catppuccin `blue`, and now I'm using the brand color (default mauve) as you suggested.
- Warn: traditionally gray. I was using catppuccin `overlay2`, and now I made it darker by using catppuccin `surface0`, as you suggested.
- Reject: I am using catppuccin `red`

Here is the updated screenshot of the same area that you showed, with the warn button darker (the red button is not an "accept" button, so it is red and not in the brand mauve color):
![botao-jupyter-warn](https://github.com/catppuccin/catppuccin/assets/244890/f4996adb-5e18-426f-8599-52f40af20d60)

Here is an example of a screen using the "accept" button, which earlier was blue, and now I use the brand color:
![botao-juypter-ok](https://github.com/catppuccin/catppuccin/assets/244890/c9de9a5e-2706-4959-b5dd-37ea46467bda)

What do you think now?
</details>

<details>
<summary>when using a dark theme variant, dropdowns in the settings menu use a light background</summary>
Unfortunately this seems to be an upstream bug in JupyterLab. 

Funny thing is that I did not notice it before you mentioned because I was using in Firefox, and it renders the list with dark background. In Chrome I was able to see the white background.

As you mentioned, even the default dark theme has this behavior, and I also tested other dark JupyterLab themes ([darcula](https://github.com/telamonian/theme-darcula), [neon-theme](https://github.com/yeebc/jupyterlab-neon-theme), [simpledark](https://github.com/ericmiguel/jupyterlab-simpledark)) and they all have this same ugly appearance :(. 

These are the related JupyterLab issues that I have found: 
 - https://github.com/jupyterlab/jupyterlab/issues/12159 
 - https://github.com/jupyterlab/jupyterlab/issues/11988
</details>

<details>
<summary>dropdowns using flat colours</summary>
I was not able to find how to specify this color with theme variables in the Jupyter Code. 
Maybe it is possible, but I was not able to find out.

I looked at other JupyterLab themes ([darcula](https://github.com/telamonian/theme-darcula), [neon-theme](https://github.com/yeebc/jupyterlab-neon-theme), [simpledark](https://github.com/ericmiguel/jupyterlab-simpledark)), and they all seem to have a flat appearance like ours. Not that it confirms that it is not possible, but at least is an evidence that it is _probably_ not possible.

In the style variables we do have these variables to set:
```
 --jp-input-box-shadow
 --jp-input-active-background
 --jp-input-hover-background
 --jp-input-background
 --jp-input-border-color
 --jp-input-active-border-color
```

I tried setting them differently, but it looks like they have no effect on he input fields.

I am not familiar with the internal JupyterLab code, but I was able to find a probable location that should be fixed: https://github.com/jupyterlab/jupyterlab/blob/696751e7cfbaca0d60801cef323282562b26fa1f/packages/ui-components/style/rjsfTemplates.css#L24. The `background` is set as none, but maybe this should be set as ` --jp-input-background` instead. Just a wild guess. Maybe I should register an issue to them about it.
</details>

&nbsp;

So, basicallly I was only able to fix the color of the buttons, making them darker. The background of the dropdown list and the color of the input field I was not able to change it, and it looks like it is not possible.

The main branch is updated, and also I released the update to PyPI as v0.1.2 if you want to test the latest change.

**Comment by backwardspy** [comment]
RE: dropdowns with light backgrounds, that is interesting as i am also on firefox!

the new buttons look good, if you are able to set the text colour independently on the accept & reject buttons, i would suggest trying `surface1` (jetbrains) or `crust` (vscode) as it may improve readability. `text` on accent colours can be hard to read depending on the chosen colour.

otherwise that's all fine, thanks again for the extra info. let me know if you can do anything with the accept/reject button text colours. i have no other comments, so once that's done we can look at transferring the port over!

**Comment by gabrielmagno** [comment]
RE: dropdown in Firefox, it seems that the dropdown in my Firefox is using my Linux Plasma theme for some reason (not sure if a KDE Plasma thing, or maybe a "Plasma Integration" extension that I use. Anyway, now I am testing both in Firefox and Chrome.

Hey @backwardspy, I have some good news.

When searching for a way to set the color of the text in the buttons, I had an idea that worked in the end. Basicallly, in the same CSS file that we define the JupyterLab color variables (which is loaded when we set a theme), I specify (override) attributes of the JupyterLab classes of the elements (e.g.: buttons). Maybe this is obvious to more experienced frontend/CSS developers, but it took me some time to have the idea because I don't have lots of experience with CSS 😅.

Then, I was able to apply this same technique to override the elements of the input fields. So we were able to fix not only the button font color issue, but also the other issues that you mentioned earlier: "dropdown list color being white" and "dropdown input fields being flat".

While doing the tests I also noticed that the "darker" color that I have put for the "reject" buttons was not good in the Latte flavor (it was actually very light and close to the background color). So I changed from `surface0` to `surface1`, and now they look good on all flavors I think.

I made some screenshots:
![buttons-input-mocha](https://github.com/catppuccin/catppuccin/assets/244890/08ff2ea0-7495-44eb-8514-dd3eda84edbc)
![buttons-input-latte](https://github.com/catppuccin/catppuccin/assets/244890/0039269e-2dcd-4cd2-9761-1612e533e9aa)

I updated the master branch and published version 0.1.3 in PyPI if you want to test it.

By the way, I was wondering that it might be relevant to mention in the README that the theme works for JupyterLab 3.x.x and 4.x.x. How do you feel about it? If that is okay to do, where would be a good place to mention it?

**Comment by backwardspy** [comment]
nice work, that looks much better. i've given it another test and it all seems to be working as expected.

regarding the note about version support, having it at the top or bottom of the usage section would be fine. it seems relevant to the installation instructions more than anything else.

i am happy for you to transfer this port into the organisation whenever you're ready. thanks again for all your hard work on this!

**Comment by backwardspy** [comment]
the next thing to potentially look at would be bringing the syntax highlighting more in line with the style guide & other ports. a lot of work has been done recently on the vscode port for this, so it might be useful as a reference.

i don't want to block accepting the port based on this though, as i assume it would be quite a large change and the port looks perfectly good as it is.

**Comment by gabrielmagno** [comment]
Thank you for the great review and feedbacks @backwardspy . I have even more respect for the catppuccin project now that I saw how careful and thoughtful you are with the ports.

Regarding the syntax highlighting, an important detail to know about JupyterLab is that it uses [CodeMirror](https://codemirror.net/) as editor. I tried looking at first to VS Code and neovim (which I had already configured and installed), but they had some differences between each other. Then I realized that catppuccin had a port to CodeMirror, then I tried using the same values as it for JupyterLab. But there was actually some minor differences, so I was in doubt. The main challenge about CodeMirror is that its parser/AST is not very detailed, so we do not have all the token types that VSCode or neovim (+treesitter) have. Some token types that have different colors in VSCode sometimes are mapped to a single token type in CodeMirror, so they would have a single color. What I tried to do was to inspire in catppuccin CodeMirror port, but then "tune" a little bit for Python and R code (considering that those are probably the most common languages utilized in JupyterLab). This "tuning" have some opinionated decisions of mine, so it might need an extra work to achieve standardization with the whole catppuccin ecosystem, or at least more opinions besides mine 😄 .

I've just transferred the repository to catppuccin organization, renamed all the repository URLs to refer the new repository URL, then released v0.2.0 with this minor changes to PyPI.

**Comment by backwardspy** [comment]
> Thank you for the great review and feedbacks @backwardspy . I have even more respect for the catppuccin project now that I saw how careful and thoughtful you are with the ports.

you are most welcome.

> Regarding the syntax highlighting (...)

that makes sense, most of our code ports have some level of variance in them due to the tokens that they provide for colouring. as i said it's not a big issue at all and the port looks very nice as it is.

thanks again for contributing your time and energy to the project, and thanks @sgoudham for adding updating the ports list!
